### PR TITLE
Replace icon fonts with Gridicons

### LINKED ISSUE #16903: UI: Webfont enhancements; drop IE8 support and implement gridicons
It looks like WC is still supporting IE8 with the [webfont file types](https://github.com/woocommerce/woocommerce/blob/master/assets/css/_fonts.scss) included. Is it time to move to [this approach](https://css-tricks.com/snippets/css/using-font-face/#article-header-id-2)? Or is IE8 still officially supported?

As an example, star.woff is 1.3kb, whereas [star.woff2](https://github.com/woocommerce/woocommerce/files/1326442/star.zip) is reduced to 0.6kb. It would be good to at least include a woff2 in the fonts files.

### PR DESCRIPTION
**Pull request by willgorham**
Closes #16903 

The objective of this PR is to replace icon fonts with Gridicons SVG icons ([repo](https://github.com/Automattic/gridicons) / [gallery](http://automattic.github.io/gridicons/)) and drop IE8 support. 

<hr/>

Lots of replacements here, so I'll add some screenshots as reference to what changed. In general, I followed the recommendations of @jameskoster here: https://github.com/woocommerce/woocommerce/issues/16903#issuecomment-331901122. There were quite a few icons not listed in that comment though. 

### Notes/considerations:

Gridicons provides a php helper function `get_gridicon()` that will return the SVG markup (located in `includes/libraries/gridicons.php`). I also dropped the raw .svg files into `assets/images/icons/gridicons/` so they can be referenced by path as well.

_I'm thinking `get_gridicon` could be added to the PHPCS exceptions for escaping content?_ Otherwise it will throw a violation when this is used.

In general I tried to put all SVGs inline instead of in pseudo-elements (`::before/::after`) primarily because as far as I know it's impossible to style the fill color of SVGs in pseudo-elements. Also relative url paths in .css files are messy, and this keeps concerns separated. The downside of this is it will break the styling of some extensions relying on core styles being in pseudo-elements, and same for themes in the few front-end-facing cases. I can switch back to some pseudo-element icons if you'd rather keep things as backward-compatible as possible.

If this is approved, I guess the next step would be dropping IE8 support in the Gruntfile?

<hr/>

The icons I *_didn't_* change are:
- The main Woo icon in the admin menu per https://github.com/woocommerce/woocommerce/issues/16903#issuecomment-331901122 (maybe a shopping cart?).
- The product list table types column (pending whatever happens to that). There are Gridicons for simple, virtual, external, and downloadable products. We'd need something for variable and grouped. **Edit: This column was removed entirely**
- The order table actions buttons. I think we could just go with text here? I think I recall a past issue about these being confusing. **Edit: These are now plain text**
- The dashboard widget. I think this is going away as well? If that's not happening soon, I could change these for now.
- The review/rating star, per https://github.com/woocommerce/woocommerce/pull/19445#issuecomment-374211584

Most notable replacements:
- ~~Review star: This will cause styling breakage for themes that define their own star CSS (like Storefront). Also, the dynamic 'add a review' stars are added by JS so are styled in a pseudo-element, which means that you can't change their color and they look different than all of the other stars. I'm not sure what the best solution here would be, or if there needs to be one.~~
- Front-end notices: Could affect custom themes as the icons are now inline and not in ::before. 
- Product/coupon metabox data tabs: Will affect extensions that add their own icons. I added support for declaring a Gridicon when adding a custom tab.
- Table sorting handles: There are a bunch of these with a lot of duplicate styling. Might be worth consolidating things into a mixin here. Some are inline and some are in pseudo elements currently.
- Remove item crosses ('X') on the layered nav / filter widgets: Could affect custom themes due to icons being inline now.

<details>
<summary>Screenshots</summary>

<details>
<summary>Admin menu</summary>
screen shot 2018-03-17 at 5 36 45 pm
</details>
<details>
<summary>Tooltip</summary>
screen shot 2018-03-17 at 5 38 34 pm<br/>screen shot 2018-03-17 at 5 39 34 pm<br/>screen shot 2018-03-17 at 5 41 00 pm
</details>
<details>
<summary>Blank states</summary>
screen shot 2018-03-17 at 6 29 26 pm
screen shot 2018-03-17 at 5 52 47 pm
screen shot 2018-03-17 at 6 01 22 pm
screen shot 2018-03-17 at 6 03 44 pm
screen shot 2018-03-17 at 6 04 13 pm
</details>
<details>
<summary>Status page check/notice/X</summary>
screen shot 2018-03-18 at 10 53 33 am
screen shot 2018-03-18 at 10 53 59 am
</details>
<details>
<summary>Products list table image header</summary>
This:<br/>
screen shot 2018-03-17 at 5 45 39 pm

Became this:  
screen shot 2018-03-17 at 5 46 08 pm
</details>
<details>
<summary>Featured product star</summary>
screen shot 2018-03-17 at 5 49 23 pm
</details>
<details>
<summary>Product meta box data tabs</summary>
screen shot 2018-03-17 at 5 50 47 pm
</details>
<details>
<summary>Coupon meta box data tabs</summary>
screen shot 2018-03-18 at 10 59 37 am
</details>
<details>
<summary>Table sorting handles</summary>
screen shot 2018-03-17 at 6 05 11 pm
<br/>screen shot 2018-03-17 at 6 08 18 pm<br/>screen shot 2018-03-17 at 6 09 21 pm
</details>
<details>
<summary>Shipping zones globe</summary>
screen shot 2018-03-17 at 6 07 07 pm
</details>
<details>
<summary>Payment gateway enabled</summary>
screen shot 2018-03-17 at 6 09 56 pm
</details>
<details>
<summary>Emails enabled/disabled/manually sent</summary>
screen shot 2018-03-17 at 6 11 25 pm
</details>
<details>
<summary>Email configure cog (removed)</summary>
Could easily go back to a cog icon if desired.
screen shot 2018-03-17 at 6 11 54 pm
</details>
<details>
<summary>Reports export button</summary>
screen shot 2018-03-17 at 6 13 10 pm
</details>
<details>
<summary>Shipping zones blank state globe</summary>
The Gridicon 'globe' is pretty weak. It looks more like a soccer ball when it's big 😄<br/>
screen shot 2018-03-17 at 6 15 14 pm
</details>
<details>
<summary>Order edit addresses</summary>
screen shot 2018-03-17 at 5 54 39 pm
</details>
<details>
<summary>Order item edit/delete</summary>
<a href="https://cl.ly/qFFE" target="_blank"></a>
</details>
<details>
<summary>Order shipping/fee/refund line item</summary>
screen shot 2018-03-17 at 5 55 02 pm
</details>
<details>
<summary>Order item image placeholder (replaced with default placeholder)</summary>
screen shot 2018-03-17 at 5 58 19 pm
</details>
<details>
<summary>Order action (replaced with text)</summary>
screen shot 2018-03-17 at 5 58 58 pm
</details>
<details>
<summary>Order preview</summary>
<a href="https://cl.ly/qGq5" target="_blank"></a>
</details>
<details>
<summary>Rating star</summary>
<a href="https://cl.ly/qFpM" target="_blank"></a><br/><a href="https://cl.ly/qFh0" target="_blank"></a>
</details>
<details>
<summary>Product variation image add/delete</summary>
<a href="https://cl.ly/qGg6" target="_blank"></a>
</details>
<details>
<summary>AJAX add to cart loading/added</summary>
<a href="https://cl.ly/qFuh" target="_blank"></a>
</details>
<details>
<summary>Notice/info/error messages</summary>
screen shot 2018-03-17 at 6 23 09 pm<br/><br/>
screen shot 2018-03-17 at 6 23 42 pm
<br/><br/>screen shot 2018-03-17 at 6 25 48 pm
</details>
<details>
<summary>Order coupon remove X</summary>
screen shot 2018-03-18 at 3 52 22 pm
</details>
<details>
<summary>My Account email/phone</summary>
screen shot 2018-03-18 at 5 04 38 pm
</details>
<details>
<summary>Layered nav, rating filter, and active filters widgets 'remove' X</summary>
screen shot 2018-03-29 at 8 25 11 pm
</details>
<details>
<summary>Product importer complete check</summary>
screen shot 2018-03-29 at 8 47 24 pm
</details>
<details>
<summary>Reports - chart widget resize handle</summary>
screen shot 2018-03-29 at 9 03 55 pm
</details>
<details>
<summary>Product downloadable files 'remove' X</summary>
screen shot 2018-03-29 at 9 19 56 pm
</details>
<details>
<summary>Admin product gallery remove X and placeholder</summary>
<a href="https://cl.ly/qYr1" target="_blank"></a>
</details>
<details>
<summary>Admin order refund amount and line item</summary>
screen shot 2018-03-30 at 12 38 48 am
screen shot 2018-03-30 at 12 46 57 am
</details>
<details>
<summary>Admin order delete tax</summary>
<a href="https://cl.ly/qYcr" target="_blank"></a>
</details>
<details>
<summary>Admin order actions</summary>
screen shot 2018-03-30 at 2 07 00 am
</details>
</details>

### CODE REVIEW

**Comment by willgorham** [comment]
Jeez, sorry about the mess of screenshots 🤦‍♂️ 
Originally I had these tied to individual commits, but in my personal repo. What's the best way to organize these for review? Happy to clean things up... 😬

File: includes/class-wc-post-types.php [hunk 1]
```diff
@@ -312,6 +312,7 @@ public static function register_post_types() {
 					'has_archive'         => $has_archive,
 					'show_in_nav_menus'   => true,
 					'show_in_rest'        => true,
+					'menu_icon'           => 'data:image/svg+xml;base64,' . base64_encode( str_replace( '<svg', '<svg fill="black"', get_gridicon( 'gridicons-product' ) ) ), // SVG 'fill' attribute must be set for menu colors to apply correctly.
```
Comment (willgorham) [inline includes/class-wc-post-types.php]: The `svg` tag here needs to have a defined `fill` attribute to get the admin menu color scheme to apply correctly. I could edit the svg directly, but the changes would get overwritten if the Gridicons library got updated. The `str_replace` isn't elegant, but it does the job.

**Comment by willgorham** [comment]
Replacing the metabox data tab icons here https://github.com/woocommerce/woocommerce/pull/19445/commits/23b0623aa2fd17afa54810c78ef315a55c2d052c is going to break some backwards compatibility with extensions that define their own tab icons (e.g. Product Bundles). But they can add a Gridicon name when defining a custom tab via `woocommerce_coupon_data_tabs` or `woocommerce_product_data_tabs`. Could go back to pseudo-element styling for compatibility if desired.

**Comment by willgorham** [comment]
https://github.com/woocommerce/woocommerce/pull/19445/commits/a6bb9047fefccf106bf969eae36c9abd1b0a3afa affects WC Subscriptions adding their Automatic Recurring Payments column to the payment gateway table (not significantly, they just lose the 'enabled' icon). Could go back to pseudo-elements to make this work for others relying on this styling as well, though they could update the column content on their end pretty easily. 

screen shot 2018-03-15 at 8 49 40 pm

**Comment by jameskoster** [comment]
> The main Woo icon in the admin menu per #16903 (comment) (maybe a shopping cart?).

For now I think we should leave this. I can provide an svg if required.

> The product list table types column (pending whatever happens to that). There are Gridicons for simple, virtual, external, and downloadable products. We'd need something for variable and grouped.

There's a vote taking place over here https://woocommerce.wordpress.com/2018/02/28/what-should-happen-to-the-product-type-column-on-the-admin-products-screen/

> The order table actions buttons. I think we could just go with text here? I think I recall a past issue about these being confusing.

Now that there's more room in the orders table we can comfortably make these text buttons.

> Review star: This will cause styling breakage for themes that define their own star CSS

We have a separate font for the star. I think we should consider leaving that as-is for now and investigate an alternative solution for things like the rating selection in the review form in a separate PR.

> Front-end notices / Archive add to cart button ajax icons: Could affect custom themes. The button icons are pseudo-elemens also, so color customization is a challenge.

They were pseudo elements before, right? So it should affect themes. Custom themes should be dequeuing the WC css anyway.

> Product/coupon metabox data tabs: Will affect extensions that add their own icons. I added support for declaring a Gridicon when adding a custom tab.

Yeah this is tricky. Ideally extensions should also be able to specify a custom .svg file for the tab as Gridicons will not cover everything.

**Comment by willgorham** [comment]
> For now I think we should leave this. I can provide an svg if required.

No need for an svg right away since the WooCommerce font is staying available for now, right? I'm happy to replace it though if you want to provide one.

> Now that there's more room in the orders table we can comfortably make these text buttons.

Sounds good - I'll take care of this.

> We have a separate font for the star. I think we should consider leaving that as-is for now

Makes sense, I'll revert the star changes.

> They were pseudo elements before, right? So it shouldn't affect themes.

Add-to-cart buttons should be fine because those are still pseudo-elements. The notice icons are inline now though, so those could be affected.

> Yeah this is tricky. Ideally extensions should also be able to specify a custom .svg file for the tab as Gridicons will not cover everything.

What do you think makes the most sense here, allowing an extension to overwrite the entire icon markup? (If that's done, then an extension could potentially add whatever image type they want, which I guess would be fine. Opens things up to more potential breakage, but I guess extensions already have plenty of those opportunities already 😄)

File: includes/class-wc-post-types.php (hunk 1 above)
Comment (willgorham) [inline includes/class-wc-post-types.php]: The `str_replace` here doesn't feel great, but the icon has to have a defined `fill` attribute for the admin menu default colors to apply correctly. It's not pretty, but this does the job.

**Comment by jameskoster** [comment]
> No need for an svg right away since the WooCommerce font is staying available for now, right?

True, we can leave for now.

> If that's done, then an extension could potentially add whatever image type they want

Yeah that's actually a good point. A mismatching icon style will negatively affect the overall UX. OK, I retract the custom icon requirement :) Dev's should be able to;

1. Do nothing and 'inherit' a default, generic tab icon
2. Select a Gridicon that fits their needs

If there's nothing currently in Gridicons suitable to their needs, we should encourage them to open an issue/PR on the Gridicons repo.

**Comment by willgorham** [comment]
Cool, makes sense to me!

I guess theoretically an extension author could hide the svg and add a pseudo-element icon with whatever they want, if they really want to go that far. (Dynamic Pricing is already kind of doing that.) But keeping it to Gridicons with a default fallback and the opportunity to request new Gridicons seems reasonable.

**Comment by willgorham** [comment]
- Reverted the star rating changes. (Maybe something to note - the 'star' font is used everywhere except the Add A Review stars, where the 'WooCommerce' font is used.)

- Updated the order action buttons to be plain text.

- I also updated some icons that I missed on the first pass:

<details>
<summary>My Account email/phone</summary>
screen shot 2018-03-18 at 5 04 38 pm
</details>
<details>
<summary>Layered nav, rating filter, and active filters widgets 'remove' X</summary>
screen shot 2018-03-29 at 8 25 11 pm
</details>
<details>
<summary>Product importer complete check</summary>
screen shot 2018-03-29 at 8 47 24 pm
</details>
<details>
<summary>Reports - chart widget resize handle</summary>
screen shot 2018-03-29 at 9 03 55 pm
</details>
<details>
<summary>Product downloadable files 'remove' X</summary>
screen shot 2018-03-29 at 9 19 56 pm
</details>
<details>
<summary>Admin product gallery remove X and placeholder</summary>
<a href="https://cl.ly/qYr1" target="_blank"></a>
</details>
<details>
<summary>Admin order refund amount and line item</summary>
screen shot 2018-03-30 at 12 38 48 am
screen shot 2018-03-30 at 12 46 57 am
</details>
<details>
<summary>Admin order delete tax</summary>
<a href="https://cl.ly/qYcr" target="_blank"></a>
</details>
<details>
<summary>Admin order actions</summary>
screen shot 2018-03-30 at 2 07 00 am
</details>
I've added these items to the main issue comment as well.

**Comment by willgorham** [comment]
Hi @mikejolley 

Sorry about the delay here, just got back from vacation. 

I totally understand about the size of this PR, half way through I realized it was going to be a lot bigger than I initially thought 😬. Figured I'd just submit it and break it up if needed - so I'll work on splitting this into more reasonable chunks. 

Thanks!

**Comment by willgorham** [comment]
Is this still relevant with wc-admin coming down the pipe? If it is, I'm happy to go ahead and break this up into smaller chunks for review when I get the chance.

**Comment by mikejolley** [comment]
@willgorham wc-admin (the dashboard part) will likely ship before 4.0. Perhaps it's best to leave this for now and revisit in the future since we're unsure how long this change would actually be relevent. There are other places that require more immediate action :) Thanks!

**Comment by willgorham** [comment]
@mikejolley Sounds good to me - thanks!

**Pull request closed without merging**